src/main/java/org/spongepowered/api/event/cause/entity/damage/DamageStep.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.event.cause.entity.damage;
+
+import org.spongepowered.api.event.Cause;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalDouble;
+
+/**
+ * A step represent an operation made by the platform (vanilla and mods) or modifiers added by plugins.
+ * Steps are structured as trees where children modify the input or output of the parent step.
+ * A damage calculation is made of multiple trees of steps.
+ */
+public interface DamageStep {
+
+    /**
+     * Gets the {@link DamageStepType} of this step.
+     *
+     * @return the step type
+     */
+    DamageStepType type();
+
+    /**
+     * Gets the {@link Cause} of this step.
+     *
+     * @return The cause of this step
+     */
+    Cause cause();
+
+    /**
+     * Gets whether this step is skipped.
+     * When skipped, only the step itself and its side effects are ignored, children are still applied.
+     * A modifier willing to ignore every previous children should revert the damage to {@link #damageBeforeChildren()},
+     * or call {@link #skip} on each child.
+     *
+     * @return Whether this step is skipped
+     */
+    boolean isSkipped();
```

Comment:
Make this enum?
```java
enum SkipOptions {
    NONE,
    SELF,
    CHILDREN_ONLY,
    ALL
}
````

Reply:
I'm worried that setting skip option NONE on a child after setting skip option ALL on a parent might be ignored.

Reply:
Fair enough. Generally uncancelling is advised against because its problematic but I guess semantically its hard to explain that you might need to walk the tree upwards to figure out if the parent is cancelled.

Reply:
Actually, it is problematic to call setSkipped with CHILDREN_BEFORE or CHILDREN_AFTER on a parent because some children might already have been traversed and some not.

I think I will make setSkipped(boolean) fail silently when the node has already been traversed and add helper methods like setChildrenBeforeSkipped(boolean) etc that will just call setSkipped recursively. So nodes already traversed are not affected, and other nodes are skipped.

Reply:
Better approach would be that we implement it as two pass system. First pass filters out nodes and on the second pass the tree structure is immutable.

Reply:
Plugins may wanna know the damage to decide whether some steps should be skipped or not.

Reply:
We cannot skip a step after the damage calculation has been done. We either skip the damage calculation and the side-effects or none. For platform steps, these two are linked and cannot be separated.

Reply:
We can do it as part of the root pass, but not to the whole tree, so its not different from what you could already do.

But running the skip pass first would be the most straightforward schematically too. Want to modify the damage based on the chain? Use the modifiers.

Reply:
Imagine step B wanna skip the side-effects of step A when a certain condition depending on the damage value is met.
This would not be possible with the two pass approach.
I don't see what's wrong with the one pass approach, it is simpler and allows combining skip, damage, and side-effects without restrictions.

Reply:
> I don't see what's wrong with the one pass approach

You brought up that
> Actually, it is problematic to call setSkipped with CHILDREN_BEFORE or CHILDREN_AFTER on a parent because some children might already have been traversed and some not.

Which with two pass approach there would be no such problem.

Reply:
Yeah my plan was not to introduce a second pass, just keeping the current approach and adding helper methods

Reply:
I will put it on code so it's clearer, maybe my message wasn't describing my idea properly

Reply:
Generally these implicit behaviors or silent suppressions are not a great design approach. Someone is going to end up shooting their own leg so it would be better if we had an approach where that is not possible, hence the suggestion.

The order of the modifiers is also not guaranteed so there is a possibility that other modifications might result to different behavior depending on whatever there are other plugins present or not.

Reply:
Well, having setSkipped fail-silently was just to make things easier, I can still write helper methods without that

Reply:
If we can fail fast and document the cases where its not possible, that's okay. It would be good to document some other use cases that people might try to accomplish.

For example, they might want to emulate skip `CHILDREN_AFTER` by appending a node that skips all the following modifiers and resets the damage to the original value.